src/memory.rs

```diff
@@ -0,0 +1,41 @@
+use std::mem;
+use std::slice;
+
+#[derive(Copy, Clone)]
+#[repr(C)]
+pub struct Buffer {
+    pub ptr: *mut u8,
+    pub size: usize,
+}
+
+// this frees memory we released earlier
+#[no_mangle]
+pub extern "C" fn free_rust(buf: Buffer) {
+    unsafe {
+        let _v = Vec::from_raw_parts(buf.ptr, buf.size, buf.size);
+    }
+}
+
+pub fn read_buffer(b: &Buffer) -> Option<&'static [u8]> {
```

Comment:
As written, this function *must* be marked unsafe, because, if the `b` is freed after function is called, the result would be a danging slice. 

Perhaps you need

```
impl Buffer {
    fn as_slice(&self) -> Option<&[u8]> // note the same lifetime
}
```